### PR TITLE
chore(DTFS2-6994): remove moment from portal-api tests

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
Moment.js is deprecated & should be replaced. This PR covers separate changes to tests, without changing behaviour - see #2691. The e2e test is covered by #2690 

## Resolution :heavy_check_mark:
Replace moment() with vanilla Dates where possible and date-fns where necessary

## Miscellaneous :heavy_plus_sign:
- rename and move unit tests labelled as api tests

## Note ⚠️ 
- need to remove package dependency once both PRs are merged

### CODE REVIEW

File: portal-api/api-tests/fixtures/deal-fully-completed-issued-and-unissued-facilities.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
-const moment = require('moment');
+const { add, format, sub } = require('date-fns');
 const dealFullyCompleted = require('./deal-fully-completed');
 const CONSTANTS = require('../../src/constants');
 
+const nowDate = new Date()
```
Comment (AlexBramhill) [inline portal-api/api-tests/fixtures/deal-fully-completed-issued-and-unissued-facilities.js:5]:
```suggestion
const nowDate = Date.now()
```

File: portal-api/api-tests/v1/deals/deals-status-facilities.api-test.js [hunk 2]
```diff
@@ -630,6 +634,8 @@ describe('/v1/deals/:id/status - facilities', () => {
         return null;
       };
 
+      // here
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/deals/deals-status-facilities.api-test.js:637]: reminder of this comment being here

File: portal-api/api-tests/v1/loan/loan-change-cover-start-date-validation.api-test.js [hunk 3]
```diff
@@ -129,7 +133,7 @@ describe('/v1/deals/:id/loan/:loanId', () => {
             submissionType: 'Automatic Inclusion Notice',
             details: {
               ...newDeal.details,
-              submissionDate: moment().subtract(2, 'day').utc().valueOf()
+              submissionDate: submissionDate.valueOf(),
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/loan/loan-change-cover-start-date-validation.api-test.js:136]: You're calling `valueOf` twice

File: portal-api/src/v1/gef/controllers/validation/eligibilityCriteria.test.js [hunk 4]
```diff
@@ -2,8 +2,8 @@ const {
   getAnsweredItems,
   isAutomaticCover,
   eligibilityCriteriaStatus,
-} = require('../../../../src/v1/gef/controllers/validation/eligibilityCriteria');
-const CONSTANTS = require('../../../../src/constants');
+} = require('./eligibilityCriteria');
```
Comment (AlexBramhill) [inline portal-api/src/v1/gef/controllers/validation/eligibilityCriteria.test.js:5]: good spot

**Review by AlexBramhill — COMMENTED** [review]
can approve after the small comments -- Good job here :)

File: portal-api/api-tests/v1/bonds/bond-issue-facility-validation.api-test.js [hunk 5]
```diff
@@ -1,24 +1,27 @@
...
 const testUserCache = require('../../api-test-users');
 const { as } = require('../../api')(app);
 const { dateValidationText } = require('../../../src/v1/validation/fields/date');
-const { formattedTimestamp } = require('../../../src/v1/facility-dates/timestamp');
 const { MAKER } = require('../../../src/v1/roles/roles');
 const { DB_COLLECTIONS } = require('../../fixtures/constants');
+const { DATE_FORMATS } = require('../../../src/constants');
+
+const nowDate  = new Date();
+const todayPlus3Months1Day = add(nowDate, { months: 3, days: 1 });
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply portal-api/api-tests/v1/bonds/bond-issue-facility-validation.api-test.js:13]: Doesnt work at 276 but have moved to 171

File: portal-api/src/v1/helpers/date.ts
Comment (MarRobSoftwire) [inline portal-api/src/v1/helpers/date.ts:1]: Taken from #2691 some functions aren't actually used in this PR

File: portal-api/api-tests/v1/deals/deals-status-submitted-updates-unsubmitted-issued-facilities.api-test.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ const api = require('../../../src/v1/api');
 const externalApis = require('../../../src/external-api/api');
 const { MAKER, CHECKER } = require('../../../src/v1/roles/roles');
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/deals/deals-status-submitted-updates-unsubmitted-issued-facilities.api-test.js:12]: double line

```suggestion
```

File: portal-api/api-tests/v1/bonds/bonds-validation.api-test.js [hunk 7]
```diff
@@ -409,14 +407,59 @@ describe('/v1/deals/:id/bond', () => {
         });
 
         describe('when has some values', () => {
-          it('should return validationError', async () => {
-            const nowDate = moment();
-            const requestedCoverStartDateFields = {
-              'requestedCoverStartDate-day': moment(nowDate).format('DD'),
-              'requestedCoverStartDate-month': '',
-              'requestedCoverStartDate-year': '',
-            };
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/bonds/bonds-validation.api-test.js:410]: Free new line


```suggestion
```

File: portal-api/api-tests/v1/bonds/bonds-validation.api-test.js [hunk 8]
```diff
@@ -459,14 +502,59 @@ describe('/v1/deals/:id/bond', () => {
         });
 
         describe('when has some values', () => {
-          it('should return validationError', async () => {
-            const nowDate = moment();
-            const coverEndDateFields = {
-              'coverEndDate-day': moment(nowDate).add(1, 'day').format('DD'),
-              'coverEndDate-month': '',
-              'coverEndDate-year': '',
-            };
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/bonds/bonds-validation.api-test.js:505]: Free new line


```suggestion
```

File: portal-api/api-tests/v1/bonds/bonds-validation.api-test.js [hunk 9]
```diff
@@ -683,41 +766,89 @@ describe('/v1/deals/:id/bond', () => {
...
         });
 
         describe('when has some values', () => {
-          it('should return validationError', async () => {
-            const date = moment().add(1, 'day');
-            const conversionRateFields = {
-              'conversionRateDate-day': moment(date).format('DD'),
-              'conversionRateDate-month': '',
-              'conversionRateDate-year': '',
-            };
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/bonds/bonds-validation.api-test.js:787]: Free new line

```suggestion
```

File: portal-api/api-tests/v1/loan/loan-change-cover-start-date-validation.api-test.js [hunk 10]
```diff
@@ -385,13 +389,13 @@ describe('/v1/deals/:id/loan/:loanId', () => {
...
 
-            const todayPlus3Months1Day = moment().add(3, 'month').add(1, 'day');
             const requestedCoverStartDateFields = {
-              'requestedCoverStartDate-day': moment(todayPlus3Months1Day).format('DD'),
-              'requestedCoverStartDate-month': moment(todayPlus3Months1Day).format('MM'),
-              'requestedCoverStartDate-year': moment(todayPlus3Months1Day).format('YYYY'),
+              'requestedCoverStartDate-day': format(todayPlus3Months1Day, 'dd'),
+              'requestedCoverStartDate-month': format(todayPlus3Months1Day, 'MM'),
+              'requestedCoverStartDate-year': format(todayPlus3Months, 'yyyy'),
             };
```
Comment (AlexBramhill) [inline portal-api/api-tests/v1/loan/loan-change-cover-start-date-validation.api-test.js:397]:
```suggestion
```

File: portal-api/src/v1/helpers/date.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
+import { format, set, startOfDay } from 'date-fns';
+import { DATE_FORMATS } from '../../constants';
+
+/**
+ * @param date 
+ * @returns Date as Unix timestamp representing the number of milliseconds between this date and 1st
+ * January 1970 (UTC), stored as a string.
+ */
+export const getDateAsEpochMillisecondString = (date: Date) => date.valueOf().toString();
```
Comment (AlexBramhill) [inline portal-api/src/v1/helpers/date.ts:9]: To be the biggest pain in the world can we have tests for this please

**Review by AlexBramhill — COMMENTED** [review]
Really good - Few formatting things to address and to add some tests to date.ts, but otherwise I'm ready for this to go!

File: portal-api/src/v1/helpers/date.test.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
+import { format } from 'date-fns';
```
Comment (AlexBramhill) [inline portal-api/src/v1/helpers/date.test.ts:1]: Could you look into the analysis errors here? I didn't notice in vscode!

File: portal-api/api-tests/fixtures/deal-fully-completed.js [hunk 13]
```diff
@@ -1,6 +1,10 @@
-const moment = require('moment');
+const { format, add, sub } = require('date-fns')
```
Comment (MarRobSoftwire) [inline portal-api/api-tests/fixtures/deal-fully-completed.js:1]: This import isn't getting any codacy problems

File: portal-api/api-tests/v1/bonds/bonds.api-test.js [hunk 14]
```diff
@@ -1,4 +1,4 @@
-const moment = require('moment');
+const { format, add } = require('date-fns');
```
Comment (MarRobSoftwire) [inline portal-api/api-tests/v1/bonds/bonds.api-test.js:1]: No codacy problems

File: portal-api/unit.jest.config.js [hunk 15]
```diff
@@ -1,5 +1,6 @@
 module.exports = {
-  collectCoverageFrom: ['src/**/*.{js,}'],
+  preset: 'ts-jest/presets/js-with-babel',
```
Comment (AlexBramhill) [inline portal-api/unit.jest.config.js:2]: With the changes below -- can we remove this?

**Review by AlexBramhill — APPROVED** [review]
lgtm -- one comment that we can look at after this is merged

**Review by AlexBramhill — APPROVED** [review]

**Review by AlexBramhill — DISMISSED** [review]
lgtm

**Pull request merged**